src/libtexture/imagecache.cpp

```diff
@@ -2822,6 +2822,48 @@ ImageCacheImpl::get_image_info(ImageCacheFile* file,
 #undef ATTR_DECODE
 }
 
+bool
+ImageCacheImpl::get_image_info_type(ustring filename, int subimage,
+                                    int miplevel, ustring dataname,
+                                    TypeDesc& datatype)
+{
+    ImageCachePerThreadInfo* thread_info = get_perthread_info();
+    ImageCacheFile* file = find_file(filename, thread_info, nullptr);
+    if (!file && dataname != s_exists) {
+        error("Invalid image file \"{}\"", filename);
+        return false;
+    }
+    return get_image_info_type(file, thread_info, subimage, miplevel, dataname,
+                               datatype);
+}
+
+bool
+ImageCacheImpl::get_image_info_type(ImageCacheFile* file,
+                                    ImageCachePerThreadInfo* thread_info,
+                                    int subimage, int miplevel,
+                                    ustring dataname, TypeDesc& datatype)
+{
+    // Output the TypeDesc of a given attribute (if found). If not found
+    // we set it to UNKNOWN.
+
+    ImageCacheStatistics& stats(thread_info->m_stats);
+    ++stats.imageinfo_queries;
+    file = verify_file(file, thread_info, true);
+
+    const ImageSpec& spec(file->spec(subimage, miplevel));
+    ParamValue tmpparam;
+    const ParamValue* p = spec.find_attribute(dataname, tmpparam);
+
+    if (p) {
+        datatype = p->type();
+        return true;
+    }
+
+    // Does it make sense to set the type here, or should we just return
+    // false?
+    datatype.basetype = TypeDesc::UNKNOWN;
+    return false;
```

Comment:
I think these lines can simply be

    return spec.getattributetype(dataname);

and the return type for `get_image_info_type` can just be TypeDesc (and the `datatype` parameter to this method can go away). The returned TypeDesc will already be UNKNOWN if it's not found.

This matches the way we've already done `getattributetype()` in ImageSpec, and also in ParamList.

Also, the ImageSpec::getattributetype works even if it's one of the "built-in" fields, and isn't restricted to the "extra" attributes.

Reply:
I take back that last sentence partially -- ImageSpec::getattributetype will only work for things that are in the ImageSpec (extra_args but also some other hard fields). But what this implementation of get_image_info_type won't be able to do is retrieve any of the *other* kinds of queries that get_image_info knows about that don't correspond to the ImageSpec, unless you are inclined to replicate much of that logic here.